crates/block-producer/src/tx_pool/mod.rs

```diff
@@ -0,0 +1,514 @@
+#![allow(unused)]
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    fmt::Display,
+    sync::Arc,
+};
+
+use miden_objects::{
+    accounts::AccountId,
+    transaction::{ProvenTransaction, TransactionId},
+    Digest,
+};
+
+use crate::store::TransactionInputs;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BatchId(u64);
+
+impl Display for BatchId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BatchId {
+    pub fn increment(mut self) {
+        self.0 += 1;
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BlockNumber(u32);
+
+impl Display for BlockNumber {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BlockNumber {
+    pub fn increment(mut self) {
+        self.0 += 1;
+    }
+}
+
+pub struct TransactionPool {
+    /// The latest inflight state of each account.
+    ///
+    /// Accounts without inflight transactions are not stored.
+    account_state: BTreeMap<AccountId, (Digest, TransactionId)>,
+
+    /// Block number at which transaction inputs are considered stale.
+    ///
+    /// This means we track inflight data and in blocks completed after this block,
+    /// **excluding** this block.
+    stale_block: BlockNumber,
+
+    /// The number of the next block we hand out.
+    next_block: BlockNumber,
+
+    /// All transactions currently inflight.
+    ///
+    /// This includes those not yet processed, those in batches and blocks after the stale block.
+    tx_pool: BTreeMap<TransactionId, InflightTransaction>,
+
+    /// Transactions ready to be included in a batch.
+    ///
+    /// aka transactions whose parent transactions are already included in batches.
+    tx_roots: BTreeSet<TransactionId>,
```

Comment:
I would maybe encapsulate these in a separate struct - something like `InFlightTransactions`.

Reply:
I've create graph types to encapsulate these.